### PR TITLE
7517 email user on important updates

### LINKED ISSUE #7517: Send email notifications when important user account information is changed
When important user data is changed, we should send an email notification to tell the user what changed. This will alert the user in the event that their account has been taken over by someone else (to be clear, this has not ever happened).

For example:
- Adding email address
- Removing email address
- Changing password
- Adding an access token

Since auth is owned by @sourcegraph/core-services I am assigning @tsenart. This is not a high priority since there is no private code on sourcegraph.com and on-premise instances generally use some sort of SSO.

### PR DESCRIPTION
**Pull request by ElizabethStirling**
Fixes #7517

If emails are configured, the user will be sent an email when important account information is changed. This currently encompasses changing/resetting the password, adding/removing emails, and adding/removing access tokens.

### CODE REVIEW

File: cmd/frontend/graphqlbackend/access_tokens.go [hunk 1]
```diff
@@ -72,6 +73,13 @@ func (r *schemaResolver) CreateAccessToken(ctx context.Context, args *createAcce
 	}
 
 	id, token, err := db.AccessTokens.Create(ctx, userID, args.Scopes, args.Note, actor.FromContext(ctx).UID)
+
+	if conf.CanSendEmail() {
+		if err := backend.UserEmails.SendUserEmailOnFieldUpdate(ctx, userID, "created an access token"); err != nil {
+			log15.Warn("Failed send email to inform user of access token creation", "error", err, ctx)
```
Comment (ryanslade) [inline cmd/frontend/graphqlbackend/access_tokens.go:79]:
```suggestion
			log15.Warn("Failed to send email to inform user of access token creation", "error", err)
```

File: cmd/frontend/graphqlbackend/access_tokens.go [hunk 2]
```diff
@@ -124,6 +132,12 @@ func (r *schemaResolver) DeleteAccessToken(ctx context.Context, args *deleteAcce
 		}
 	}
 
+	if conf.CanSendEmail() {
+		if err := backend.UserEmails.SendUserEmailOnFieldUpdate(ctx, token.SubjectUserID, "deleted an access token"); err != nil {
+			log15.Warn("Failed send email to inform user of access token deletion", "error", err, ctx)
```
Comment (ryanslade) [inline cmd/frontend/graphqlbackend/access_tokens.go:137]:
```suggestion
			log15.Warn("Failed to send email to inform user of access token deletion", "error", err)
```

File: cmd/frontend/graphqlbackend/user.go [hunk 3]
```diff
@@ -309,6 +310,12 @@ func (r *schemaResolver) UpdatePassword(ctx context.Context, args *struct {
 	if err := db.Users.UpdatePassword(ctx, user.ID, args.OldPassword, args.NewPassword); err != nil {
 		return nil, err
 	}
+
+	if conf.CanSendEmail() {
+		if err := backend.UserEmails.SendUserEmailOnFieldUpdate(ctx, user.ID, "updated the password"); err != nil {
+			log15.Warn("Failed send email to inform user of password update", "error", err, ctx)
```
Comment (ryanslade) [inline cmd/frontend/graphqlbackend/user.go:316]:
```suggestion
			log15.Warn("Failed to send email to inform user of password update", "error", err)
```

File: cmd/frontend/graphqlbackend/user_emails.go [hunk 4]
```diff
@@ -75,6 +75,13 @@ func (r *schemaResolver) AddUserEmail(ctx context.Context, args *struct {
 	if err := backend.UserEmails.Add(ctx, userID, args.Email); err != nil {
 		return nil, err
 	}
+
+	if conf.CanSendEmail() {
+		if err := backend.UserEmails.SendUserEmailOnFieldUpdate(ctx, userID, "added an email"); err != nil {
+			log15.Warn("Failed send email to inform user of email addition", "error", err, ctx)
```
Comment (ryanslade) [inline cmd/frontend/graphqlbackend/user_emails.go:81]:
```suggestion
			log15.Warn("Failed to send email to inform user of email addition", "error", err)
```

File: cmd/frontend/graphqlbackend/user_emails.go [hunk 5]
```diff
@@ -101,6 +108,12 @@ func (r *schemaResolver) RemoveUserEmail(ctx context.Context, args *struct {
 		return nil, err
 	}
 
+	if conf.CanSendEmail() {
+		if err := backend.UserEmails.SendUserEmailOnFieldUpdate(ctx, userID, "removed an email"); err != nil {
+			log15.Warn("Failed send email to inform user of email removal", "error", err, ctx)
```
Comment (ryanslade) [inline cmd/frontend/graphqlbackend/user_emails.go:113]:
```suggestion
			log15.Warn("Failed to send email to inform user of email removal", "error", err)
```

File: cmd/frontend/internal/auth/userpasswd/reset_password.go [hunk 6]
```diff
@@ -186,6 +187,12 @@ func HandleResetPasswordCode(w http.ResponseWriter, r *http.Request) {
 		http.Error(w, "Password reset code was invalid or expired.", http.StatusUnauthorized)
 		return
 	}
+
+	if conf.CanSendEmail() {
+		if err := backend.UserEmails.SendUserEmailOnFieldUpdate(ctx, params.UserID, "reset the password"); err != nil {
+			log15.Warn("Failed send email to inform user of password reset", "error", err, ctx)
```
Comment (ryanslade) [inline cmd/frontend/internal/auth/userpasswd/reset_password.go:193]:
```suggestion
			log15.Warn("Failed to send email to inform user of password reset", "error", err)
```

**Review by ryanslade — APPROVED** [review]
LGTM apart from the suggested changes. 

Note that it's generally not that useful to log the context as it they're usually only used to allow cancellations and timeouts to cascade though the system.

File: cmd/frontend/backend/user_emails.go [hunk 7]
```diff
@@ -181,3 +181,48 @@ Verify your email address {{printf "%q" .Email}} on Sourcegraph by following thi
 <p><strong><a href="{{.URL}}">Verify email address</a></p>
 `,
 })
+
+// SendUserEmailOnFieldUpdate sends the user an email that important account information has changed.
+// The change is the information we want to provide the user about the change
+func (userEmails) SendUserEmailOnFieldUpdate(ctx context.Context, id int32, change string) error {
+	email, _, err := db.UserEmails.GetPrimaryEmail(ctx, id)
+	if err != nil {
+		log15.Warn("Failed to get user email", "error", err, ctx)
```
Comment (unknwon) [inline cmd/frontend/backend/user_emails.go:190]: Is there a special reason we want to log the error here while we propagate (and potentially somewhere up in the call chain logs this error again)?

**Pull request merged**